File: Backend/src/models/task.model.js
import mongoose, { Schema } from 'mongoose';

const taskSchema = new Schema({
  // 1. ADDED: Title (Crucial for the Admin Dashboard list view)
  title: {
    type: String,
    required: [true, 'A short title is required'],
    trim: true
  },
  rawReportText: {
    type: String,
    required: [true, 'Original report text is required']
  },
  category: {
    type: String,
    required: true,
    enum: [
      'Medical', 'Rescue', 'Food & Water', 'Shelter', 
      'Sanitation', 'Labor', 'Transport', 'Supplies', 
      'Animal Rescue', 'Infrastructure', 'Other'
    ]
  },
  severity: {
    type: Number,
    required: true,
    min: 1,
    max: 5
  },
  requiredSkills: {
    type: [String],
    default: []
  },
  // 2. ADDED: Human-readable location (Crucial for Volunteers who don't just use GPS)
  locationDescription: {
    type: String,
    required: [true, 'A textual description of the location is required'],
    trim: true
  },
  location: {
    type: {
      type: String,
      enum: ['Point'],
      default: 'Point'
    },
    coordinates: {
      type: [Number], // [Longitude, Latitude]
      required: true
    }
  },
  status: {
    type: String,
    enum: ['Pending', 'Matched', 'In Progress', 'Completed', 'Cancelled'],
    default: 'Pending'
  },
  completionNote: {
    type: String,
    trim: true,
    default: ""
  },
  reportedBy: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true 
  },
  assignedVolunteer: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    default: null 
  }
}, { 
  timestamps: true 
});

taskSchema.index({ location: '2dsphere' });

// This makes the dashboard sorting lightning fast
taskSchema.index({ status: 1, severity: -1, createdAt: -1 });

export const Task = mongoose.model('Task', taskSchema);

File: frontend_app/lib/core/models/task_model.dart
class TaskModel {
  final String? id;
  final String title;
  final String category;
  final String rawReportText;
  final int severity;
  final List<String> requiredSkills;
  final String locationDescription;
  final String status;
  final String? completionNote;
  final List<double>? locationCoordinates; // [lng, lat]

  TaskModel({
    this.id,
    required this.title,
    required this.category,
    required this.rawReportText,
    required this.severity,
    required this.requiredSkills,
    required this.locationDescription,
    this.status = 'Pending',
    this.completionNote,
    this.locationCoordinates,
  });

  factory TaskModel.fromJson(Map<String, dynamic> json) {
    return TaskModel(
      id: json['_id'],
      title: json['title'] ?? '',
      category: json['category'] ?? 'Other',
      rawReportText: json['rawReportText'] ?? '',
      severity: int.tryParse(json['severity']?.toString() ?? '') ?? 3,
      requiredSkills: json['requiredSkills'] != null ? List<String>.from(json['requiredSkills']) : [],
      locationDescription: json['locationDescription'] ?? '',
      status: json['status'] ?? 'Pending',
      completionNote: json['completionNote'],
      locationCoordinates: json['location'] != null && json['location']['coordinates'] != null
          ? List<double>.from((json['location']['coordinates'] as List).map((e) => (e as num).toDouble()))
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'title': title,
      'category': category,
      'rawReportText': rawReportText,
      'severity': severity,
      'requiredSkills': requiredSkills,
      'locationDescription': locationDescription,
      'status': status,
      if (completionNote != null) 'completionNote': completionNote,
      if (locationCoordinates != null) 'location': {
        'type': 'Point',
        'coordinates': locationCoordinates,
      }
    };
  }
}
